Create Proposed Constitution

jackkoenig opened the pull request.

jackkoenig committed:
Create Proposed Constitution

jackkoenig committed:
Add Technical Advisory Committee and references to CHIPS

jackkoenig committed:
Some cleanups and rewording

jackkoenig commented:
TODO:
- [ ] Mark as provisional (since we haven't officially joined CHIPS Alliance)
- [ ] Apply minor revisions from June 6th governance meeting

azidar committed:
Update constitution.md

azidar committed:
Update constitution.md

azidar committed:
Added tac.md

azidar committed:
Update tac.md

azidar committed:
Update tac.md

chick reviewed (dismissed):
Thing of beauty

ucbjrl reviewed (dismissed):
lgtm

seldridge reviewed (dismissed):
Signed

edwardcwang commented on constitution.md:
```diff
@@ -0,0 +1,232 @@
+Reviewers strictly subsume Developers (e.g. they also have all the rights and responsibilities of Developers).
+
+### Developers
+Developers are the most common membership type, and is an entry point for enthusiastic community members. They have voting
```
Has this ever happened / has this happened yet?

azidar replied:
No, primarily because we haven't had a constitution yet. Do you have an objection to this idea?

edwardcwang commented on constitution.md:
```diff
@@ -0,0 +1,232 @@
+Developers are the most common membership type, and is an entry point for enthusiastic community members. They have voting
+rights for high-level directives as set by the administrators of a given project.
+
+Anyone can become a developer by signing up and *signing the CLA*. In exchange for voting rights, a developer must commit to 
```
Especially given this

edwardcwang commented on constitution.md:
```suggestion
The CWG’s standard communication channels and development policies are outlined as follows, and will likely be adopted for other 
```

edwardcwang commented on constitution.md:
```diff
@@ -0,0 +1,232 @@
+the CHIPS Alliance Board.
+
+## Mission Statement
+The Chisel ecosystem is a stable, user-friendly platform for production-quality RTL generator development, 
```
```suggestion
The Chisel ecosystem is intended to provide a stable, user-friendly platform for production-quality RTL generator development, 
```

grebe replied:
I don't think this language needs to be weakened.

azidar replied:
Is this satisfactory to leave, @edwardcwang ?

edwardcwang replied:
My personal take is that this comes across a little bit as self-congratulatory - e.g. "mission accomplished, we are the best". Though it's also fine with me to leave it as-is.

edwardcwang commented on constitution.md:
```diff
@@ -0,0 +1,232 @@
+The TAC is made up of members of the community as well as one member appointed by
+the CHIPS Alliance Board.
+The members are listed by name and Github username in a document within
+this repository. Membership change is reflected by pull requests
```
```suggestion
this repository. Membership changes are reflected by pull requests
```

edwardcwang commented on constitution.md:
Should we really hardcode Gitter into the constitution? I was thinking almost leaving it as a chat room / channel (e.g. especially with our experiment with matrix etc)

seldridge replied:
Probably not. This is likely worth revising.

azidar replied:
Agreed. Weird to have this in the constitution. Let's remove all explicit communication app mentions and instead have 'a commitment to communication by a variety of means`?

edwardcwang commented on constitution.md:
```suggestion
Generally, mature projects will release minor/major versions, where minor versions are backwards compatible, 
```

edwardcwang commented on constitution.md:
```suggestion
and major versions can break old APIs. Old APIs will be deprecated prior to removal.
```

I assume ok = old?

edwardcwang commented on constitution.md:
While I am a 100% CI superfan, sometimes CI does introduce issues (gets hung up, some yosys check fails for silly reasons, etc). Maybe we can add a clause (if it's not already elsewhere) to add a backup provision for this?

seldridge replied:
Maybe just "pass testing". We already have the ability to merge PRs that rightly fail tests (`[skip chisel tests]` in FIRRTL). Also, this does technically enshrine `master` branch into the constitution which implies a specific development model. Maybe we can generalize this?

I think, broadly, this is trying to say that:

1. We have a culture of testing
1. Projects should have reviewers
1. People should not be committing directly to a release branch
1. These rules *may* not apply if you have a very immature project

azidar replied:
Replace with 'pass continuous integration tests' with 'rigorous testing' ?

edwardcwang commented on constitution.md:
To avoid issues where X group feels like they are not included in the special protected list from Y-ism this could be shortened to

```
In the interest of fostering an open and welcoming environment, we as contributors and administrators pledge to make participation in our project and our community a harassment-free experience for everyone.
```

edwardcwang commented on constitution.md:
Possibility to add a clause indicating that the overall goal of this CoC is to advance the technical development of the Chisel ecosystem. CoCs have been controversial in some projects where it has derailed technical development

azidar replied:
After some discussion, let's just take an existing code of conduct, which has been vetted, verbatim.

edwardcwang reviewed (dismissed):
Generally good, left nitpicks/food for thoughts below.

schoeberl commented:
This sounds all good to me. What do I need to do to approve, except leaving this positive comment?

Martin

ducky64 commented on constitution.md:
```diff
@@ -0,0 +1,232 @@
+responsible for upholding the policies outlined in this document.
+
+Other responsibilities include: admitting (and if necessary expelling) projects,
+organizing community meetings/conferences, resolving disputes between projects,
```
perhaps a nit, but also resolving disputes within projects, were the community can't come to a consensus

ducky64 commented on constitution.md:
```diff
@@ -0,0 +1,232 @@
+allocating funds, clarifying procedures for each membership type, and publicizing
+the Chisel community.
+
+Projects are admitted (or expelled) by a 2/3 vote of the TAC.
```
Clarification: 2/3 of all TAC members (so abstain counts as a no vote), or 2/3 of TAC members who cast a vote (so abstaining doesn't affect the vote proportion either way)? Or more broadly, is abstaining allowed, and is that the default for anyone who doesn't vote?

azidar replied:
@seldridge will do some research and plan is to discuss at next meeting.

ducky64 commented on constitution.md:
I think it should also be allowed (perhaps even encouraged) for the community to nominate new Reviewers (in addition to being able to self-nominate).

ducky64 reviewed (commented):
This mostly looks fine, but a few minor comments looking through the document. I also share @edwardcwang's concern about hardcoding communication channel implementation details here - while StackOverflow and GitHub probably won't go away anytime soon, chat room platform seems up in the air.

schoeberl commented:
Looks like am too dumb for GitHub. I see that I have been added to this repo. But I do not see how I can approve this PR.

Martin

schoeberl reviewed (dismissed):
LGT

Finally managed how to do this ;-)

grebe commented on constitution.md:
```diff
@@ -0,0 +1,232 @@
+Other responsibilities include: admitting (and if necessary expelling) projects,
+organizing community meetings/conferences, resolving disputes between projects,
+allocating funds, clarifying procedures for each membership type, and publicizing
+the Chisel community.
```
Should delegation be explicitly called out as a responsibility? For example, it could be the TAC's role to delegate some of it's responsibilities to a sub-committee they select. I suppose delegation is more of an ability than it is a responsibility, but it seems like something that would be good to mention somewhere.

azidar replied:
I think delegation is implicit in any leadership position.

grebe commented on constitution.md:
Very minor nit, but I think it is incorrect to use an apostrophe with an initialism. `s/PR's/PRs/`. Other examples below.

grebe commented on constitution.md:
Should we explicitly reference semantic versioning?

azidar replied:
Ok to be high level in the constitution, as changes to versioning likely shouldn't require a change to the constitution.

grebe commented on constitution.md:
I haven't seen bodysize as one word before. Should it be two?

azidar committed:
Grammar

Co-Authored-By: edwardcwang <edwardcwang@users.noreply.github.com>

azidar committed:
Update constitution.md

Co-Authored-By: edwardcwang <edwardcwang@users.noreply.github.com>

azidar committed:
Update constitution.md

Co-Authored-By: edwardcwang <edwardcwang@users.noreply.github.com>

azidar committed:
Update constitution.md

Co-Authored-By: edwardcwang <edwardcwang@users.noreply.github.com>

azidar commented on constitution.md:
```suggestion
participation in our project and our community a harassment-free experience for everyone, regardless of age, body size, 
```

azidar committed:
Apply suggestions from code review

azidar commented on constitution.md:
```suggestion
All code contributions to mature projects (Chisel, FIRRTL, etc.) must undergo a review process and pass rigorous testing prior to merging into the master branch.
```

azidar committed:
Update constitution.md

azidar commented on constitution.md:
```diff
@@ -0,0 +1,230 @@
+responsible for upholding the policies outlined in this document.
+
+Other responsibilities include: admitting (and if necessary expelling) projects,
+organizing community meetings/conferences, resolving disputes between projects,
```
```suggestion
organizing community meetings/conferences, resolving disputes within and between projects,
```

azidar committed:
Update constitution.md

azidar commented on constitution.md:
```suggestion
had over 10 reviews on a project, the individual can nominate themselves, or the community can nominate them, to become a Reviewer. A request is then sent to the administrators, who decide whether to add the individual as a Reviewer for that project.
```

azidar commented on constitution.md:
```suggestion
contribute through PRs to Chisel repositories, and are encouraged to review and comment but they cannot accept PR requests. It is hoped that volunteers will become developers.
```

azidar commented on constitution.md:
```suggestion
The CWGs standard communication channels and development policies are outlined as follows, and will likely be adopted for other 
```

azidar commented on constitution.md:
```suggestion
New features which are experimental could be released in an **experimental** package, whereby all APIs in this package are 
```

azidar commented on constitution.md:
```suggestion
User-facing support is primarily through a commitment to answering questions, providing documentation, and promoting community dialogue via a variety of communication channels listed on the website.

```

azidar commented on constitution.md:
```suggestion

All Contributors are expected to abide by the code of conduct detailed in `CODE_OF_CONDUCT.md`, which is a verbatim copy of [Contributor Covenant 2.0](https://www.contributor-covenant.org/version/2/0/code_of_conduct/code_of_conduct.md).

```

edwardcwang replied:
I think similar concerns apply to using this verbatim though in this case.

azidar committed:
Update small grammar, clarifications, and remove code of conduct section.

azidar committed:
Add code_of_conduct.md

azidar commented on constitution.md:
```diff
@@ -0,0 +1,179 @@
+allocating funds, clarifying procedures for each membership type, and publicizing
+the Chisel community.
+
+Projects are admitted (or expelled) by a 2/3 vote of the TAC.
```
```suggestion
Projects are admitted (or expelled) by a 2/3 vote of the TAC, assuming a quorum is present.
```

azidar commented on constitution.md:
```diff
@@ -0,0 +1,179 @@
+Chair elections generally occur during scheduled meetings, but in the event of an
+absent or otherwise misbehaving chair, such a vote can be initiated by any TAC member
+as a PR against the TAC membership document.
+
```
```suggestion

#### Voting Procedure
Voting can only proceed if a quorum of 2/3 of all TAC members is established. An abstained vote counts towards quorum but not towards determining whether a motion passes or fails. Voting rules and procedure is enforced by the TAC Chair.
```

azidar committed:
Update voting/abstaining procedure

azidar commented:
Ratification of Constitution! If you want to ratify, please approve this PR. We will ratify once all TAC members approve.

chick reviewed (dismissed):
Nice work

ducky64 commented on constitution.md:
```diff
@@ -0,0 +1,181 @@
+Voting can only proceed if a quorum of 2/3 of all TAC members is established. An abstained vote counts towards quorum but not towards determining whether a motion passes or fails. Voting rules and procedure is enforced by the TAC Chair.
+### Administrator
+Consists of one lead and any number of additional administrators who direct a
+given project and report progress to the TAC
```
Missing period at end of sentence

azidar replied:
```suggestion
given project and report progress to the TAC.
```

ducky64 reviewed (dismissed):
lgtm, found one typo

azidar committed:
Add missing period

azidar commented:
Had to add a missing period. Please re-verify!

jackkoenig commented:
LGTM (I can't officially approve because I made the PR)

chick reviewed (dismissed):
Let's go

edwardcwang commented on code_of_conduct.md:
```diff
@@ -0,0 +1,130 @@
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be
+reported to the community leaders responsible for enforcement at
+[INSERT CONTACT METHOD].
```
Was this intended to be left as a placeholder?

azidar replied:
```suggestion
chisel.language@gmail.com.
```

edwardcwang commented on code_of_conduct.md:
```diff
@@ -0,0 +1,130 @@
+
+## Scope
+
+This Code of Conduct applies within all community spaces, and also applies when
```
I think this should make it clear that it does not and should not include conduct unrelated to the Chisel project. For example, commentary on non-official social media or private e-mails should not be considered.

edwardcwang commented on code_of_conduct.md:
```diff
@@ -0,0 +1,130 @@
+
+## Our Pledge
+
+We as members, contributors, and leaders pledge to make participation in our
```
I think this may almost be better and more generically inclusive if just left at "We as members, contributors, and leaders pledge to make participation in our community a harassment-free experience for everyone", given the intent of the paragraph.

azidar replied:
We want to keep the standard contributor covenant code of conduct, as we didn't write and it provides more of a standard.

azidar replied:
I personally like being explicit about this. Are there examples that you can think of which is not included?

edwardcwang commented on code_of_conduct.md:
```diff
@@ -0,0 +1,130 @@
+Instances of abusive, harassing, or otherwise unacceptable behavior may be
+reported to the community leaders responsible for enforcement at
+[INSERT CONTACT METHOD].
+All complaints will be reviewed and investigated promptly and fairly.
```
Who will investigate? Will it be this committee? Based on the language in the main document ("resolving disputes within and between projects") it seems like this is the case.

seldridge committed:
Add CoC language clarifying the scope

This clarifies the language of the Code of Conduct to affirmatively
state the implied inverse of the scope. This adds language stating
that activities outside of the community are out of scope and have no
bearing on being in good standing with the CoC.

Signed-off-by: Schuyler Eldridge <schuyler.eldridge@ibm.com>

seldridge committed:
Clarify process for CoC investigations

Signed-off-by: Schuyler Eldridge <schuyler.eldridge@ibm.com>

seldridge committed:
Merge pull request #4 from freechipsproject/coc-language

Update CoC with Scoping Language, Clarification on Investigation Committee Procedure

edwardcwang committed:
Update code_of_conduct.md

seldridge committed:
Merge pull request #5 from freechipsproject/patch-1

Update code_of_conduct.md

chick reviewed (approved):
Looks good to me

jackkoenig commented:
LGTM (again). I can't officially approve because I made the PR.

jackbackrack reviewed (approved):
LGTM

azidar commented:
Merging!
